### PR TITLE
Added LinearMemory trait

### PR DESCRIPTION
**Pull request by ptitSeb**
Added LinearMemory trait, to describe memory used in runtime.

### CODE REVIEW

File: lib/api/src/sys/externals/memory.rs [hunk 1]
```diff
@@ -60,6 +60,13 @@ impl Memory {
         })
     }
 
+    /// Create a memory object from an existing memory and attaches it to the store
+    pub fn new_from_existing(new_store: &mut impl AsStoreMut, memory: VMMemory) -> Self {
```
Comment (syrusakbary) [inline lib/api/src/sys/externals/memory.rs:64]: Can this be `Into<VMMemory>`? That way, any implementation that implements the `LinearMemory` trait will work

File: lib/types/src/memory.rs [hunk 2]
```diff
@@ -113,10 +121,89 @@ unsafe impl MemorySize for Memory64 {
...
     fn offset_to_native(offset: Self::Offset) -> Self::Native {
         offset as Self::Native
     }
     fn native_to_offset(native: Self::Native) -> Self::Offset {
         native as Self::Offset
     }
 }
+
+/// Represents memory that is used by the WebAsssembly module
+pub trait LinearMemory
```
Comment (syrusakbary) [inline lib/types/src/memory.rs:134]: Can we move this, along with the `VMMemoryDefinition` back to the `wasmer-vm` crate?
The JS implementation will not make use of it (and can't make use of it), so the VM seems a much better place

File: lib/api/src/js/externals/memory.rs [hunk 3]
```diff
@@ -116,6 +100,21 @@ impl Memory {
...
+        store: &mut impl AsStoreMut,
+        js_memory: js_sys::WebAssembly::Memory,
+        ty: MemoryType,
+    ) -> Result<Self, MemoryError> {
+        let vm_memory = VMMemory::new(js_memory, ty);
+        Ok(Self::from_vm_export(store, vm_memory))
+    }
+
+    /// Create a memory object from an existing memory and attaches it to the store
+    pub fn new_from_existing(new_store: &mut impl AsStoreMut, memory: VMMemory) -> Self {
```
Comment (syrusakbary) [inline lib/api/src/js/externals/memory.rs:114]: `Into<VMMemory>` to support the proposal for sys

File: lib/api/src/sys/externals/memory.rs (hunk 1 above)
Comment (ptitSeb) [reply lib/api/src/sys/externals/memory.rs:64]: but about the store parameter?

**Comment by syrusakbary** [comment]
@ptitSeb can you add tests using a custom memory implementation?

**Comment by syrusakbary** [comment]
@ptitSeb I think we want the custom memory to be used from the Tunables to make sure the wasmer API is well-rounded so people can use it properly

File: lib/api/src/sys/tunables.rs [hunk 4]
```diff
@@ -174,4 +174,138 @@ mod tests {
...
+            Ok(VMMemory::from_custom(memory))
+        }
+        unsafe fn create_vm_memory(
+            &self,
+            _ty: &MemoryType,
+            _style: &MemoryStyle,
+            _vm_definition_location: NonNull<VMMemoryDefinition>,
+        ) -> Result<VMMemory, MemoryError> {
+            let memory = VMTinyMemory::new().unwrap();
+            Ok(VMMemory::from_custom(memory))
```
Comment (syrusakbary) [inline lib/api/src/sys/tunables.rs:268]: Can we do `Ok(memory.into())` (as I believe the `From` is implemented)

File: lib/api/src/sys/tunables.rs [hunk 5]
```diff
@@ -174,4 +174,138 @@ mod tests {
...
+        fn table_style(&self, _table: &TableType) -> TableStyle {
+            TableStyle::CallerChecksSignature
+        }
+        fn create_host_memory(
+            &self,
+            _ty: &MemoryType,
+            _style: &MemoryStyle,
+        ) -> Result<VMMemory, MemoryError> {
+            let memory = VMTinyMemory::new().unwrap();
+            Ok(VMMemory::from_custom(memory))
```
Comment (syrusakbary) [inline lib/api/src/sys/tunables.rs:259]: Can we do `Ok(memory.into())` (as I believe the `From` is implemented)

File: lib/vm/src/vmcontext.rs [hunk 6]
```diff
@@ -303,120 +303,77 @@ mod test_vmglobal_import {
     }
 }
 
-/// The fields compiled code needs to access to utilize a WebAssembly linear
```
Comment (syrusakbary) [inline lib/vm/src/vmcontext.rs:306]: We should probably keep `VMMemoryDefinition` inside of the `vmcontext`

Comment (ptitSeb) [reply lib/vm/src/vmcontext.rs:306]: done

**Review by syrusakbary — APPROVED** [review]

**Pull request merged**